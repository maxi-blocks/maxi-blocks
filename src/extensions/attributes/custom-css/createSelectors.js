const PSEUDO_ELEMENTS = ['before', 'after'];

const createSelectors = (rawSelectors, addPseudoElementSelectors = true) => {
	const getNormalAndHoverSelectors = ({ label, target }) => {
		const pseudoElement = PSEUDO_ELEMENTS.find(pseudo =>
			label.includes(pseudo)
		);
		const targetWithoutPseudoElement = target.replace(
			`::${pseudoElement}`,
			''
		);

		return {
			normal: {
				label,
				target,
			},
			hover: {
				label: `${label} on hover`,
				target: `${targetWithoutPseudoElement}:hover${
					pseudoElement ? `::${pseudoElement}` : ''
				}`,
			},
		};
	};

	const addPseudoElementSelector = (
		key,
		{ target, label },
		pseudoElement,
		obj
	) => {
		obj[`${pseudoElement} ${key}`] = {
			label: `${label} ::${pseudoElement}`,
			target: `${target}::${pseudoElement}`,
		};
	};

	const selectors = Object.entries(rawSelectors).reduce(
		(acc, [label, target]) => {
			acc[label] = {
				label,
				target,
			};
			return acc;
		},
		{}
	);

	const result = { ...selectors };

	if (addPseudoElementSelectors)
		PSEUDO_ELEMENTS.forEach(pseudoElement => {
			Object.entries(selectors).forEach(([key, selector]) =>
				addPseudoElementSelector(key, selector, pseudoElement, result)
			);
		});

	Object.entries(result).forEach(([key, selector]) => {
		result[key] = getNormalAndHoverSelectors(selector);
	});

	return result;
};

export default createSelectors;
