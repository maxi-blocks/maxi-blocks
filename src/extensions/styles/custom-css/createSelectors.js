const createSelectors = (selectors, addPseudoElementSelectors = true) => {
	const getNormalAndHoverSelectors = obj => {
		const { label, target } = obj;

		return {
			normal: obj,
			hover: {
				label: `${label} on hover`,
				target: `${target}:hover`,
			},
		};
	};

	const addPseudoElementSelector = (key, selector, pseudoElement, obj) => {
		const { label, target } = selector;

		obj[`${pseudoElement} ${key}`] = {
			label: `${label} ::${pseudoElement}`,
			target: `${target}::${pseudoElement}`,
		};
	};

	const result = { ...selectors };

	if (addPseudoElementSelectors)
		['before', 'after'].forEach(pseudoElement => {
			Object.entries(selectors).forEach(([key, selector]) => {
				addPseudoElementSelector(key, selector, pseudoElement, result);
			});
		});

	Object.entries(result).forEach(([key, obj]) => {
		result[key] = getNormalAndHoverSelectors(obj);
	});

	return result;
};

export default createSelectors;
