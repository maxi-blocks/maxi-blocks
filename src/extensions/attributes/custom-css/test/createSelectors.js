import createSelectors from '../createSelectors';

describe('createSelectors', () => {
	const selectors = {
		canvas: '',
		button: 'buttonClassName',
	};

	it('Should create selectors', () => {
		expect(createSelectors(selectors)).toMatchSnapshot();
	});

	it('Should create selectors without pseudo elements', () => {
		expect(createSelectors(selectors, false)).toMatchSnapshot();
	});
});
