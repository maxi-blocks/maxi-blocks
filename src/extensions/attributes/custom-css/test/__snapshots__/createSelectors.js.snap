// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createSelectors Should create selectors 1`] = `
Object {
  "after button": Object {
    "hover": Object {
      "label": "button ::after on hover",
      "target": "buttonClassName:hover::after",
    },
    "normal": Object {
      "label": "button ::after",
      "target": "buttonClassName::after",
    },
  },
  "after canvas": Object {
    "hover": Object {
      "label": "canvas ::after on hover",
      "target": ":hover::after",
    },
    "normal": Object {
      "label": "canvas ::after",
      "target": "::after",
    },
  },
  "before button": Object {
    "hover": Object {
      "label": "button ::before on hover",
      "target": "buttonClassName:hover::before",
    },
    "normal": Object {
      "label": "button ::before",
      "target": "buttonClassName::before",
    },
  },
  "before canvas": Object {
    "hover": Object {
      "label": "canvas ::before on hover",
      "target": ":hover::before",
    },
    "normal": Object {
      "label": "canvas ::before",
      "target": "::before",
    },
  },
  "button": Object {
    "hover": Object {
      "label": "button on hover",
      "target": "buttonClassName:hover",
    },
    "normal": Object {
      "label": "button",
      "target": "buttonClassName",
    },
  },
  "canvas": Object {
    "hover": Object {
      "label": "canvas on hover",
      "target": ":hover",
    },
    "normal": Object {
      "label": "canvas",
      "target": "",
    },
  },
}
`;

exports[`createSelectors Should create selectors without pseudo elements 1`] = `
Object {
  "button": Object {
    "hover": Object {
      "label": "button on hover",
      "target": "buttonClassName:hover",
    },
    "normal": Object {
      "label": "button",
      "target": "buttonClassName",
    },
  },
  "canvas": Object {
    "hover": Object {
      "label": "canvas on hover",
      "target": ":hover",
    },
    "normal": Object {
      "label": "canvas",
      "target": "",
    },
  },
}
`;
